src/Chapter3.hs

```diff
@@ -480,6 +517,55 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle
+    = NoCastle
+    | Castle String
+    | CastleWithWall String
+    deriving (Show)
+
+data EduBuilding
+    = Church
+    | Library
+    deriving (Show)
+
+data House
+    = OnePerson
+    | TwoPeople
+    | ThreePeople
+    | FourPeople
+    deriving (Show)
+
+data MagicalCity = MkCity
+    { cityCastle   :: Castle
+    , cityBuilding :: EduBuilding
+    , cityHouses   :: [House]
+    } deriving (Show)
+
+buildCastle :: MagicalCity -> String -> MagicalCity
+buildCastle city castleName = case cityCastle city of
+    CastleWithWall _ -> city { cityCastle = CastleWithWall castleName }
+    _ -> city { cityCastle = Castle castleName }
+
+buildHouse :: MagicalCity -> House -> MagicalCity
+buildHouse city house = city { cityHouses = house : cityHouses city }
+
+buildWalls :: MagicalCity -> MagicalCity
+buildWalls city = case cityCastle city of
+    Castle name -> if countCitizens (cityHouses city) >= 10
+      then city { cityCastle = CastleWithWall name } else city
+        where
+        countCitizensInHouse :: House -> Int
+        countCitizensInHouse OnePerson = 1
+        countCitizensInHouse TwoPeople = 2
+        countCitizensInHouse ThreePeople = 3
+        countCitizensInHouse FourPeople = 4
+
+        countCitizens :: [House] -> Int
+        countCitizens [] = 0
+        countCitizens (x:xs) = countCitizensInHouse x + countCitizens xs
```

Comment:
This function can be implemented without the need to pattern match on list manually by using the combination of standard functions `map` and `sum`